File: github_cpp/2/104.cpp


struct CacheEntry
{
    int val;
    int key;
    CacheEntry *prev;
    CacheEntry *next;
    CacheEntry(int v, int k) : key(k), val(v), prev(NULL), next(NULL) {}
};

class LRUCache{
public:
    LRUCache(int capacity) {
	this->capacity = capacity;
	size = 0;
	head = new CacheEntry(0, -1);
	tail = new CacheEntry(0, -1);
	head->next = tail;
	head->prev = NULL;
	tail->next = NULL;
	tail->prev = head;
    }

    ~LRUCache()
        {
	    delete head;
	    delete tail;
	}

    int get(int key) {
	CacheEntry *node = hashmap[key];
	if(node)
	{
	    detach(node);
	    attach(node);
	    return node->val;
	}
	else
	    return -1;
    }

    void set(int key, int value) {
	CacheEntry *node = hashmap[key];
	if(node)
	{
	    node->val = value;
	    detach(node);
	    attach(node);
	}
	else
	{
	    if(size >= capacity) 
	    {
		node = tail->prev;
		node->val = value;
		hashmap.erase(node->key);
		node->key = key;
		detach(node);
	    }
	    else
	    {
		++size;
		node = new CacheEntry(value, key);
	    }
	    attach(node);
	    hashmap[key] = node;
	}
    }

private:
    void detach(CacheEntry *node) {
	node->prev->next = node->next;
	node->next->prev = node->prev;
    }

    void attach(CacheEntry *node) {
	node->prev = head;
	head->next->prev = node;
	node->next = head->next;
	head->next = node;
    }

private:
    unordered_map<int, CacheEntry *> hashmap;
    int capacity;
    int size;
    CacheEntry *head;
    CacheEntry *tail;
};

File: github_code/lru-cache/java/104.java
/*
@Copyright:LintCode
@Author:   vadxin
@Problem:  http://www.lintcode.com/problem/lru-cache
@Language: Java
@Datetime: 16-08-30 07:24
*/

class Node {
    public Node pre;
    public Node next;
    public int key;
    public int val;
    public Node(int key, int value) {
        this.key = key;
        this.val = value;
        this.pre = null;
        this.next = null;
    }    
}
public class Solution {
    public Node head;
    public Node tail;
    public int capacity;
    public Map<Integer, Node> nodes;
    // @param capacity, an integer
    public Solution(int capacity) {
        // write your code here
        this.head = new Node(Integer.MAX_VALUE, Integer.MAX_VALUE);
        this.tail = new Node(Integer.MIN_VALUE, Integer.MIN_VALUE);
        head.next = tail;
        tail.pre = head;
        this.capacity = capacity;
        nodes = new HashMap<Integer, Node>();
    }

    // @return an integer
    public int get(int key) {
        // write your code here
        if (!nodes.containsKey(key)) {
            return -1;
        }
        Node cur = nodes.get(key);
        cur.pre.next = cur.next;
        cur.next.pre = cur.pre;
        move_to_tail(cur);
        return cur.val;
    }

    // @param key, an integer
    // @param value, an integer
    // @return nothing
    public void set(int key, int value) {
        // write your code here
        if (get(key) == -1) {
            Node newNode = new Node(key, value);
            if (nodes.size() >= capacity) {
                nodes.remove(head.next.key);
                head.next = head.next.next;
                head.next.pre = head;
            }
            nodes.put(key, newNode);
            move_to_tail(newNode);
        } else {
            Node cur = nodes.get(key);
            cur.val = value;
            cur.pre.next = cur.next;
            cur.next.pre = cur.pre;
            move_to_tail(cur);
        }
        return;
    }
    
    public void move_to_tail(Node cur) {
        tail.pre.next = cur;
        cur.pre = tail.pre;
        cur.next = tail;
        tail.pre = cur;
    }
}